Source/Particles/Deposition/CurrentDeposition.H

```diff
@@ -1533,6 +1533,670 @@ void doChargeConservingDepositionShapeNImplicit (const amrex::ParticleReal * con
+        cost_real = (amrex::Real *) amrex::The_Managed_Arena()->alloc(sizeof(amrex::Real));
+        *cost_real = 0._rt;
+    }
+#endif
+    amrex::ParallelFor(
+        np_to_deposit,
+        [=] AMREX_GPU_DEVICE (long const ip) {
+#if defined(WARPX_USE_GPUCLOCK)
+            const auto KernelTimer = ablastr::parallelization::KernelTimer(
+                cost && (load_balance_costs_update_algo == LoadBalanceCostsUpdateAlgo::GpuClock),
+                cost_real);
+#endif
+
+#if !defined(WARPX_DIM_3D)
+            constexpr amrex::ParticleReal inv_c2 = 1._prt/(PhysConst::c*PhysConst::c);
+
+            // Compute inverse Lorentz factor, the average of gamma at time levels n and n+1
+            // The uxp,uyp,uzp are the velocities at time level n+1/2
+            const amrex::ParticleReal uxp_np1 = 2._prt*uxp_nph[ip] - uxp_n[ip];
+            const amrex::ParticleReal uyp_np1 = 2._prt*uyp_nph[ip] - uyp_n[ip];
+            const amrex::ParticleReal uzp_np1 = 2._prt*uzp_nph[ip] - uzp_n[ip];
+            const amrex::ParticleReal gamma_n = std::sqrt(1._prt + (uxp_n[ip]*uxp_n[ip] + uyp_n[ip]*uyp_n[ip] + uzp_n[ip]*uzp_n[ip])*inv_c2);
+            const amrex::ParticleReal gamma_np1 = std::sqrt(1._prt + (uxp_np1*uxp_np1 + uyp_np1*uyp_np1 + uzp_np1*uzp_np1)*inv_c2);
+            const amrex::ParticleReal gaminv = 2.0_prt/(gamma_n + gamma_np1);
+#endif
+
+            // wqx, wqy wqz are particle current in each direction
+            Real wq = q*wp[ip];
+            if (do_ionization){
+                wq *= ion_lev[ip];
+            }
+
+            ParticleReal xp_nph, yp_nph, zp_nph;
+            GetPosition(ip, xp_nph, yp_nph, zp_nph);
+
+#if !defined(WARPX_DIM_1D_Z)
+            ParticleReal const xp_np1 = 2._prt*xp_nph - xp_n[ip];
+#else
+            ignore_unused(xp_n);
+#endif
+#if defined(WARPX_DIM_3D) || defined(WARPX_DIM_RZ)
+            ParticleReal const yp_np1 = 2._prt*yp_nph - yp_n[ip];
+#else
+            ignore_unused(yp_n);
+#endif
+            ParticleReal const zp_np1 = 2._prt*zp_nph - zp_n[ip];
+
+            // computes current and old position in grid units
+#if defined(WARPX_DIM_RZ)
+            amrex::Real const xp_new = xp_np1;
+            amrex::Real const yp_new = yp_np1;
+            amrex::Real const xp_mid = xp_nph;
+            amrex::Real const yp_mid = yp_nph;
+            amrex::Real const xp_old = xp_n[ip];
+            amrex::Real const yp_old = yp_n[ip];
+            amrex::Real const rp_new = std::sqrt(xp_new*xp_new + yp_new*yp_new);
+            amrex::Real const rp_old = std::sqrt(xp_old*xp_old + yp_old*yp_old);
+            amrex::Real const rp_mid = (rp_new + rp_old)/2._rt;
+            amrex::Real costheta_mid, sintheta_mid;
+            if (rp_mid > 0._rt) {
+                costheta_mid = xp_mid/rp_mid;
+                sintheta_mid = yp_mid/rp_mid;
+            } else {
+                costheta_mid = 1._rt;
+                sintheta_mid = 0._rt;
```

Comment:
Confused by the variable name: does this use a small angle approximation for the `sin()`?

Reply:
This was copied from Dave Grote's implicit deposition routine. The appended "_mid" for variable names means it is at time n + 1/2. There is no small angle approximation made. If you are confused about why costheta_mid and sintheta_mid are defined as they are, it is because this definition is needed to keep exact charge conservation. It will be described in a forthcoming paper, but I can share with you now.

Reply:
Thanks, ok! Yes happy to read that up.